{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "\n",
    "# Introduction\n",
    "\n",
    "---\n",
    "Up to this point, we’ve covered the foundational concepts of machine learning. You may have already experimented with various hyperparameters to optimize model performance. In this notebook, you’ll be introduced to several advanced techniques designed to further enhance your models.\n",
    "\n",
    "\n",
    "This notebook consists the following parts:\n",
    "\n",
    "- [A: Data retrieval ](#01)\n",
    "- [B: Feature engineering](#02)\n",
    "- [C: Model evaluation](#03)\n",
    "- [D: Ensemble learning](#04)\n",
    "- [E: Pipelines](#05)\n",
    "- [F: Bring it all together](#06)\n",
    "- [G: Bonus: ML OPS](#07)\n",
    "\n",
    "\n",
    "---\n",
    "\n",
    "### Learning Objectives\n",
    "By the end of this two weeks you will be able to:\n",
    "- Understand the fundamental concepts of ensemble learning.\n",
    "- Use evaluation techniques to assess models performance.\n",
    "- Enhance model performance by feature engineering. \n",
    "- Build pipelines for model development and preprocessing\n",
    "\n",
    "---\n",
    "\n",
    "### Instructions\n",
    "- Ensure you fully understand the requirements and objectives of the assignment.\n",
    "- Review the notebooks refered in the tasks\n",
    "- If you need additional context or clarification, please check the provided videos or background literature.\n",
    "- Work through each part of the assignment methodically, ensuring all tasks are completed.\n",
    "- Update your repository with your new created work\n",
    "\n",
    "### Additional Notes:\n",
    "- Do not add datafiles to your repository. Repositories with datafiles will not be accepted\n",
    "- Class solutions should be delivered in python files. Not in notebooks\n",
    "- When AI tools are used, you must provide proper references and explanations for how they were utilized. Failure to do so will be considered as academic fraud\n",
    "- The bonus assignment are not mandatory\n",
    "- Use PEP8 \n",
    "\n",
    "Good luck!\n",
    "\n",
    "F.Feenstra\n",
    "\n",
    "---\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "<a name='01'></a>\n",
    "## Part A. Data retrieval\n",
    "\n",
    "The dataset you can use for this notebook is the lung dataset from Maastricht University. It comprises 89 non-small cell lung cancer (NSCLC) patients records who underwent surgical treatment. The study where the data is from explored the relationship between radiomic imaging features and gene expression profiles. The samples were collected through biopsies at MAASTRO Clinic in The Netherlands, and the dataset is publicly available.\n",
    "\n",
    "The authors of the related paper discovered that a prognostic radiomic signature, which captures intra-tumor heterogeneity, is closely associated with underlying gene expression patterns. Developing a machine learning model to predict histology from the Clinical and Genetic Lung data can improve diagnostic accuracy and treatment personalization. In this notebook we will develop such a prediction model. \n",
    "\n",
    "**Availabel Datasets**:\n",
    "- Lung metadata dataset [1]\n",
    "- Gene expression dataset [2]\n",
    "\n",
    "**Important**\n",
    "<span style=\"background-color: lightgreen;color: black\">It is also allowed to use your own dataset from your own project if this data is highly dimensional and contains genetic information.</span>\n",
    "\n",
    "[1] [NSCLC-Radiomics-Genomics](https://wiki.cancerimagingarchive.net/display/Public/NSCLC-Radiomics-Genomics#16056856db10d39adf704eefa53e41edcf5ef41c)\n",
    "\n",
    "[2] [Gene Expression Data - GSE58661](https://ftp.ncbi.nlm.nih.gov/geo/series/GSE58nnn/GSE58661/matrix/)\n",
    "\n",
    "[3] Aerts HJWL, Rios Velazquez E, Leijenaar RTH, Parmar C, Grossmann P, Carvalho S, Bussink J, Monshouwer R, Haibe-Kains B, Rietveld D, Hoebers F, Rietbergen MM, Leemans CR, Dekker A, Quackenbush J, Gillies RJ, & Lambin P. (2015). Data From NSCLC-Radiomics-Genomics. The Cancer Imaging Archive. https://doi.org/10.7937/K9/TCIA.2015.L4FRET6Z\n",
    "\n",
    "\n",
    "### <span style=\"background-color: lightyellow;\">Data retrieval task</span>\n",
    "- Retrieve the data. (No cleaning needed yet)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Index(['sample.name', 'title', 'CEL.file', 'source.location', 'organism',\n",
      "       'characteristics.tag.gender', 'characteristics.tag.histology',\n",
      "       'characteristics.tag.tumor.size.maximumdiameter',\n",
      "       'characteristics.tag.stage.primary.tumor',\n",
      "       'characteristics.tag.stage.nodes', 'characteristics.tag.stage.mets',\n",
      "       'characteristics.tag.primaryVSmets', 'characteristics.tag.grade',\n",
      "       'molecule tested', 'label', 'platform'],\n",
      "      dtype='object')\n",
      "Index(['sample.name', 'title', 'CEL.file', 'source.location', 'organism',\n",
      "       'gender', 'histology', 'tumor.size.maximumdiameter',\n",
      "       'stage.primary.tumor', 'stage.nodes', 'stage.mets', 'primaryVSmets',\n",
      "       'grade', 'molecule tested', 'label', 'platform'],\n",
      "      dtype='object')\n",
      "Index(['title', 'histology', 'tumor.size.maximumdiameter',\n",
      "       'stage.primary.tumor', 'gender'],\n",
      "      dtype='object')\n",
      "    title                       histology  tumor.size.maximumdiameter  \\\n",
      "0  lung_1    Squamous Cell Carcinoma, NOS                         4.0   \n",
      "1  lung_2  Adenocarcinoma, Papillary, NOS                         1.3   \n",
      "2  lung_3                  Non-Small Cell                        11.0   \n",
      "4  lung_5    Squamous Cell Carcinoma, NOS                         7.8   \n",
      "5  lung_6             Adenocarcinoma, NOS                         3.5   \n",
      "\n",
      "  stage.primary.tumor gender  \n",
      "0                 pT2      M  \n",
      "1                 pT1      M  \n",
      "2                 pT3      M  \n",
      "4                 pT3      F  \n",
      "5                 pT2      M  \n",
      "Index(['GSM1416528', 'GSM1416529', 'GSM1416530', 'GSM1416531', 'GSM1416532',\n",
      "       'GSM1416533', 'GSM1416534', 'GSM1416535', 'GSM1416536', 'GSM1416537',\n",
      "       'GSM1416538', 'GSM1416539', 'GSM1416540', 'GSM1416541', 'GSM1416542',\n",
      "       'GSM1416543', 'GSM1416544', 'GSM1416545', 'GSM1416546', 'GSM1416547',\n",
      "       'GSM1416548', 'GSM1416549', 'GSM1416550', 'GSM1416551', 'GSM1416552',\n",
      "       'GSM1416553', 'GSM1416554', 'GSM1416555', 'GSM1416556', 'GSM1416557',\n",
      "       'GSM1416558', 'GSM1416559', 'GSM1416560', 'GSM1416561', 'GSM1416562',\n",
      "       'GSM1416563', 'GSM1416564', 'GSM1416565', 'GSM1416566', 'GSM1416567',\n",
      "       'GSM1416568', 'GSM1416569', 'GSM1416570', 'GSM1416571', 'GSM1416572',\n",
      "       'GSM1416573', 'GSM1416574', 'GSM1416575', 'GSM1416576', 'GSM1416577',\n",
      "       'GSM1416578', 'GSM1416579', 'GSM1416580', 'GSM1416581', 'GSM1416582',\n",
      "       'GSM1416583', 'GSM1416584', 'GSM1416585', 'GSM1416586', 'GSM1416587',\n",
      "       'GSM1416588', 'GSM1416589', 'GSM1416590', 'GSM1416591', 'GSM1416592',\n",
      "       'GSM1416593', 'GSM1416594', 'GSM1416595', 'GSM1416596', 'GSM1416597',\n",
      "       'GSM1416598', 'GSM1416599', 'GSM1416600', 'GSM1416601', 'GSM1416602',\n",
      "       'GSM1416603', 'GSM1416604', 'GSM1416605', 'GSM1416606', 'GSM1416607',\n",
      "       'GSM1416608', 'GSM1416609', 'GSM1416610', 'GSM1416611', 'GSM1416612',\n",
      "       'GSM1416613', 'GSM1416614', 'GSM1416615', 'GSM1416616'],\n",
      "      dtype='object')\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>GSM1416528</th>\n",
       "      <th>GSM1416529</th>\n",
       "      <th>GSM1416530</th>\n",
       "      <th>GSM1416531</th>\n",
       "      <th>GSM1416532</th>\n",
       "      <th>GSM1416533</th>\n",
       "      <th>GSM1416534</th>\n",
       "      <th>GSM1416535</th>\n",
       "      <th>GSM1416536</th>\n",
       "      <th>GSM1416537</th>\n",
       "      <th>...</th>\n",
       "      <th>GSM1416607</th>\n",
       "      <th>GSM1416608</th>\n",
       "      <th>GSM1416609</th>\n",
       "      <th>GSM1416610</th>\n",
       "      <th>GSM1416611</th>\n",
       "      <th>GSM1416612</th>\n",
       "      <th>GSM1416613</th>\n",
       "      <th>GSM1416614</th>\n",
       "      <th>GSM1416615</th>\n",
       "      <th>GSM1416616</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ID_REF</th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>AFFX-BioB-3_at</th>\n",
       "      <td>7.376915</td>\n",
       "      <td>8.024915</td>\n",
       "      <td>7.522543</td>\n",
       "      <td>7.152864</td>\n",
       "      <td>7.211031</td>\n",
       "      <td>7.704962</td>\n",
       "      <td>7.779374</td>\n",
       "      <td>7.136893</td>\n",
       "      <td>7.160221</td>\n",
       "      <td>8.141578</td>\n",
       "      <td>...</td>\n",
       "      <td>7.611992</td>\n",
       "      <td>7.345206</td>\n",
       "      <td>7.316975</td>\n",
       "      <td>7.418340</td>\n",
       "      <td>7.756286</td>\n",
       "      <td>7.215856</td>\n",
       "      <td>7.465109</td>\n",
       "      <td>7.245458</td>\n",
       "      <td>7.039592</td>\n",
       "      <td>7.393667</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>AFFX-BioB-5_at</th>\n",
       "      <td>6.984530</td>\n",
       "      <td>7.427048</td>\n",
       "      <td>7.077207</td>\n",
       "      <td>6.849513</td>\n",
       "      <td>6.753131</td>\n",
       "      <td>7.195947</td>\n",
       "      <td>7.429800</td>\n",
       "      <td>6.745630</td>\n",
       "      <td>6.751358</td>\n",
       "      <td>7.632631</td>\n",
       "      <td>...</td>\n",
       "      <td>7.175015</td>\n",
       "      <td>7.014479</td>\n",
       "      <td>6.913879</td>\n",
       "      <td>6.898449</td>\n",
       "      <td>7.347570</td>\n",
       "      <td>6.740992</td>\n",
       "      <td>7.080787</td>\n",
       "      <td>6.725812</td>\n",
       "      <td>6.575376</td>\n",
       "      <td>7.041106</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>AFFX-BioB-M_at</th>\n",
       "      <td>7.330576</td>\n",
       "      <td>8.010530</td>\n",
       "      <td>7.334551</td>\n",
       "      <td>7.143286</td>\n",
       "      <td>7.077163</td>\n",
       "      <td>7.753484</td>\n",
       "      <td>7.712429</td>\n",
       "      <td>7.036303</td>\n",
       "      <td>7.089186</td>\n",
       "      <td>7.992361</td>\n",
       "      <td>...</td>\n",
       "      <td>7.612393</td>\n",
       "      <td>7.285821</td>\n",
       "      <td>7.324635</td>\n",
       "      <td>7.408880</td>\n",
       "      <td>7.745624</td>\n",
       "      <td>7.152722</td>\n",
       "      <td>7.422373</td>\n",
       "      <td>7.263596</td>\n",
       "      <td>6.975162</td>\n",
       "      <td>7.366620</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>AFFX-BioC-3_at</th>\n",
       "      <td>10.922741</td>\n",
       "      <td>11.390638</td>\n",
       "      <td>10.936703</td>\n",
       "      <td>10.791909</td>\n",
       "      <td>10.701328</td>\n",
       "      <td>11.221732</td>\n",
       "      <td>11.084229</td>\n",
       "      <td>10.588372</td>\n",
       "      <td>10.628909</td>\n",
       "      <td>11.412668</td>\n",
       "      <td>...</td>\n",
       "      <td>11.097486</td>\n",
       "      <td>10.908682</td>\n",
       "      <td>10.781228</td>\n",
       "      <td>10.935723</td>\n",
       "      <td>11.200635</td>\n",
       "      <td>10.707980</td>\n",
       "      <td>10.736531</td>\n",
       "      <td>10.770440</td>\n",
       "      <td>10.367009</td>\n",
       "      <td>10.753221</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>AFFX-BioC-5_at</th>\n",
       "      <td>11.032030</td>\n",
       "      <td>11.533338</td>\n",
       "      <td>11.018510</td>\n",
       "      <td>10.875259</td>\n",
       "      <td>10.823792</td>\n",
       "      <td>11.276973</td>\n",
       "      <td>11.178643</td>\n",
       "      <td>10.700420</td>\n",
       "      <td>10.779332</td>\n",
       "      <td>11.470336</td>\n",
       "      <td>...</td>\n",
       "      <td>11.265739</td>\n",
       "      <td>10.990373</td>\n",
       "      <td>10.900765</td>\n",
       "      <td>11.101546</td>\n",
       "      <td>11.343329</td>\n",
       "      <td>10.842697</td>\n",
       "      <td>10.900509</td>\n",
       "      <td>10.838008</td>\n",
       "      <td>10.494806</td>\n",
       "      <td>10.863600</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>5 rows × 89 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                GSM1416528  GSM1416529  GSM1416530  GSM1416531  GSM1416532  \\\n",
       "ID_REF                                                                       \n",
       "AFFX-BioB-3_at    7.376915    8.024915    7.522543    7.152864    7.211031   \n",
       "AFFX-BioB-5_at    6.984530    7.427048    7.077207    6.849513    6.753131   \n",
       "AFFX-BioB-M_at    7.330576    8.010530    7.334551    7.143286    7.077163   \n",
       "AFFX-BioC-3_at   10.922741   11.390638   10.936703   10.791909   10.701328   \n",
       "AFFX-BioC-5_at   11.032030   11.533338   11.018510   10.875259   10.823792   \n",
       "\n",
       "                GSM1416533  GSM1416534  GSM1416535  GSM1416536  GSM1416537  \\\n",
       "ID_REF                                                                       \n",
       "AFFX-BioB-3_at    7.704962    7.779374    7.136893    7.160221    8.141578   \n",
       "AFFX-BioB-5_at    7.195947    7.429800    6.745630    6.751358    7.632631   \n",
       "AFFX-BioB-M_at    7.753484    7.712429    7.036303    7.089186    7.992361   \n",
       "AFFX-BioC-3_at   11.221732   11.084229   10.588372   10.628909   11.412668   \n",
       "AFFX-BioC-5_at   11.276973   11.178643   10.700420   10.779332   11.470336   \n",
       "\n",
       "                ...  GSM1416607  GSM1416608  GSM1416609  GSM1416610  \\\n",
       "ID_REF          ...                                                   \n",
       "AFFX-BioB-3_at  ...    7.611992    7.345206    7.316975    7.418340   \n",
       "AFFX-BioB-5_at  ...    7.175015    7.014479    6.913879    6.898449   \n",
       "AFFX-BioB-M_at  ...    7.612393    7.285821    7.324635    7.408880   \n",
       "AFFX-BioC-3_at  ...   11.097486   10.908682   10.781228   10.935723   \n",
       "AFFX-BioC-5_at  ...   11.265739   10.990373   10.900765   11.101546   \n",
       "\n",
       "                GSM1416611  GSM1416612  GSM1416613  GSM1416614  GSM1416615  \\\n",
       "ID_REF                                                                       \n",
       "AFFX-BioB-3_at    7.756286    7.215856    7.465109    7.245458    7.039592   \n",
       "AFFX-BioB-5_at    7.347570    6.740992    7.080787    6.725812    6.575376   \n",
       "AFFX-BioB-M_at    7.745624    7.152722    7.422373    7.263596    6.975162   \n",
       "AFFX-BioC-3_at   11.200635   10.707980   10.736531   10.770440   10.367009   \n",
       "AFFX-BioC-5_at   11.343329   10.842697   10.900509   10.838008   10.494806   \n",
       "\n",
       "                GSM1416616  \n",
       "ID_REF                      \n",
       "AFFX-BioB-3_at    7.393667  \n",
       "AFFX-BioB-5_at    7.041106  \n",
       "AFFX-BioB-M_at    7.366620  \n",
       "AFFX-BioC-3_at   10.753221  \n",
       "AFFX-BioC-5_at   10.863600  \n",
       "\n",
       "[5 rows x 89 columns]"
      ]
     },
     "execution_count": 1,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "import pandas as pd\n",
    "import gzip\n",
    "\n",
    "lung_pd = pd.read_excel('Lung3.metadata.xls')\n",
    "print(lung_pd.columns)\n",
    "lung_pd.rename(columns=lambda x: x.replace('characteristics.tag.', ''), inplace=True)\n",
    "print(lung_pd.columns)\n",
    "cols = ['title', 'histology', 'tumor.size.maximumdiameter', 'stage.primary.tumor', 'gender']\n",
    "lung_pd = lung_pd[cols]\n",
    "lung_pd.dropna(inplace=True)\n",
    "print(lung_pd.columns)\n",
    "print(lung_pd.head())\n",
    "\n",
    "lung_pd['histology'] = lung_pd['histology'].apply(lambda x: x.replace('\\xa0', ' ').lower().strip())\n",
    "uniques = set(list(lung_pd['histology']))\n",
    "\n",
    "\n",
    "data_gse = pd.read_csv('GSE58661_series_matrix.txt.gz', sep='\\t', comment='!', quotechar='\"', compression='gzip')\n",
    "data_gse.dropna(inplace=True)\n",
    "data_gse.set_index(\"ID_REF\", inplace=True)\n",
    "print(data_gse.columns)\n",
    "data_gse.head()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "---\n",
    "<a name='02'></a>\n",
    "## Part B. Feature engineering\n",
    "\n",
    "Mind you that choosing an algorithm and hyperparameter tuning might not be enough. If your data is of low quality, the algorithm will have poor performance as well. This is where feature engineering comes into play. Feature engineering involves transforming raw data into meaningful features that better represent the underlying problem to the predictive models, ultimately enhancing the model's performance. By selecting, creating, and refining features, you ensure that your data highlights the most relevant patterns and relationships, allowing the algorithm to learn more effectively. Proper feature engineering can often make the difference between a moderate model and a highly accurate one, even more so than the choice of algorithm itself. \n",
    "Possible modifications:\n",
    "- creation of new features derived from original features\n",
    "- selection of features\n",
    "- encoding features\n",
    "- log transformation \n",
    "- scaling\n",
    "- dimension reduction (*e.g.* PCA)\n",
    "\n",
    "### <span style=\"background-color: lightyellow;\">Feature engineering Task</span>\n",
    "- Review the [Study Case Feature Engineering notebook](../Study_Cases/study_case_feature_engineering.ipynb).\n",
    "- Review the [Study Case RNA-seq Preparation notebook](../Study_Cases/study_case_scanpy_object.ipynb).\n",
    "- Assess which data preparation and feature engineering steps could be beneficial for your dataset.\n",
    "- Implement a `DataProcessor` class tailored to your dataset and test it in the cell below. **Make sure it's an appropriate sklearn object, such as a `Transformer` or `Pipeline`.**\n",
    "- Update your repository with a new directory named `optimization`, including the following:\n",
    "    - The `DataProcessor` class as a Python file, complete with thorough documentation.\n",
    "    - An evaluation document that details and justifies your choices using a well-reasoned, argumentative approach.\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "N component to reach 50% explained of the variance: 16\n",
      "Explained variance list: [0.09358163 0.08701857 0.04763505 0.04399918 0.03330774 0.02687047\n",
      " 0.02465432 0.02241215 0.01850155 0.01716716 0.01618323 0.01520436\n",
      " 0.01432665 0.01350128 0.01334884 0.01303402]\n",
      "    title                       histology  tumor.size.maximumdiameter  \\\n",
      "0  lung_1    squamous cell carcinoma, nos                         4.0   \n",
      "1  lung_2  adenocarcinoma, papillary, nos                         1.3   \n",
      "2  lung_3                  non-small cell                        11.0   \n",
      "4  lung_5    squamous cell carcinoma, nos                         7.8   \n",
      "5  lung_6             adenocarcinoma, nos                         3.5   \n",
      "\n",
      "  stage.primary.tumor gender      PCA_1       PCA_2      PCA_3      PCA_4  \\\n",
      "0                 pT2      M -12.563557  143.351731 -74.116869 -29.403301   \n",
      "1                 pT1      M -14.590991 -136.112014 -31.727267 -93.328784   \n",
      "2                 pT3      M -72.357969   31.079279  -7.229756 -23.025768   \n",
      "4                 pT3      F -61.900053   -2.744339 -34.229472  42.636141   \n",
      "5                 pT2      M  17.563956  -86.546824 -10.003734  41.721048   \n",
      "\n",
      "       PCA_5  ...      PCA_7      PCA_8      PCA_9     PCA_10      PCA_11  \\\n",
      "0 -22.700421  ...  32.931793 -34.379329  64.755027 -12.162316   14.496096   \n",
      "1 -35.660518  ...  13.965305   5.555669  -4.355312  62.485834   55.839800   \n",
      "2  -3.130893  ... -45.429426 -78.416742  87.901543  -4.988843  113.737775   \n",
      "4  44.386074  ... -51.083975   0.727167 -17.386705 -26.271695  -31.518746   \n",
      "5  17.739587  ... -24.559155   0.611563  25.484848 -16.669331   12.329431   \n",
      "\n",
      "      PCA_12     PCA_13     PCA_14     PCA_15     PCA_16  \n",
      "0   0.106897   5.006088  11.083352 -24.381862 -35.678259  \n",
      "1 -34.731741   0.847731  74.123503   9.795867 -28.029201  \n",
      "2 -23.674288 -27.823004  -5.515901 -63.109644  64.606572  \n",
      "4 -23.971975 -19.356858  32.893979 -15.448110  15.009304  \n",
      "5  19.712823  -3.650841  11.197979  -9.736696  17.744318  \n",
      "\n",
      "[5 rows x 21 columns]\n"
     ]
    }
   ],
   "source": [
    "# YOUR CODE HERE TO DEMONSTRATE THE USAGE OF THE PREPROCESSOR CLASS\n",
    "from sklearn.decomposition import PCA\n",
    "from sklearn.preprocessing import StandardScaler\n",
    "import numpy as np\n",
    "\n",
    "data_gse_trans = data_gse.transpose()\n",
    "\n",
    "# To combat skewed data \n",
    "data_gse_log = np.log1p(data_gse_trans)\n",
    "\n",
    "#Standardizeing since genes can have very different scales of expression.\n",
    "scaler = StandardScaler()\n",
    "scaled_gse_trans_log = scaler.fit_transform(data_gse_log)\n",
    "\n",
    "# Performing PCA I know there is a better way to do like doing a pca on all data and then look at what point it becomes more than 0.5 total. \n",
    "# But this is to practice recursion.\n",
    "def find_components(data, start=2):\n",
    "    pca = PCA(n_components=start)\n",
    "    pca_res = pca.fit_transform(data)\n",
    "    expl_var_list = pca.explained_variance_ratio_\n",
    "\n",
    "    if sum(expl_var_list) >= 0.5:\n",
    "        print(f\"N component to reach 50% explained of the variance: {start}\")\n",
    "        print(f\"Explained variance list: {expl_var_list}\")\n",
    "        return start, pca, expl_var_list, pca_res\n",
    "    else:\n",
    "        start += 1\n",
    "        return find_components(data, start)\n",
    "\n",
    "\n",
    "start, pca, expl_var_list, pca_res = find_components(scaled_gse_trans_log)\n",
    "columns1 = [f\"PCA_{i+1}\" for i in range(start)]\n",
    "\n",
    "\n",
    "pca_res_df = pd.DataFrame(data=pca_res, columns=columns1) \n",
    "\n",
    "merged_df = pd.merge(lung_pd, pca_res_df, left_index=True, right_index=True)\n",
    "print(merged_df.head())"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Encoder used for turning strings/classifications into numeric values\n",
    "from sklearn.preprocessing import LabelEncoder\n",
    "\n",
    "def encode_column(df, column_enc):\n",
    "    encoder = LabelEncoder()\n",
    "    df[column_enc] = encoder.fit_transform(df[column_enc])\n",
    "    return df\n",
    "\n",
    "merged_df = encode_column(merged_df, 'histology')\n",
    "merged_df = encode_column(merged_df, 'stage.primary.tumor')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "N component to reach 50% explained of the variance: 14\n",
      "Explained variance list: [0.13625197 0.0657839  0.04725546 0.04508164 0.03637917 0.02773846\n",
      " 0.02481676 0.02130975 0.02095438 0.01704936 0.01577596 0.0155997\n",
      " 0.01448788 0.01410956]\n",
      "['title', 'gender', 'histology', 'stage.primary.tumor']\n",
      "    title gender                       histology  tumor.size.maximumdiameter  \\\n",
      "0  lung_1      M    Squamous Cell Carcinoma, NOS                         4.0   \n",
      "1  lung_2      M  Adenocarcinoma, Papillary, NOS                         1.3   \n",
      "2  lung_3      M                  Non-Small Cell                        11.0   \n",
      "4  lung_5      F    Squamous Cell Carcinoma, NOS                         7.8   \n",
      "5  lung_6      M             Adenocarcinoma, NOS                         3.5   \n",
      "\n",
      "  stage.primary.tumor      PCA_1     PCA_2     PCA_3      PCA_4     PCA_5  \\\n",
      "0                 pT2  15.075251 -5.472487  0.142375   2.999663  3.193272   \n",
      "1                 pT1  -6.797374  9.108923 -4.152204  10.089548 -3.865868   \n",
      "2                 pT3   7.656481  5.642052  2.395748   0.510114 -0.126368   \n",
      "4                 pT3   6.930521 -1.980997 -3.089379  -2.864192 -4.976604   \n",
      "5                 pT2  -6.866991  0.156104 -4.951233  -4.216752  0.500523   \n",
      "\n",
      "      PCA_6     PCA_7     PCA_8     PCA_9    PCA_10    PCA_11    PCA_12  \\\n",
      "0 -0.846291 -8.426318  2.013082  0.514004 -1.496055 -5.192893  0.891502   \n",
      "1  2.240864 -0.691223  2.967306 -2.303766 -1.854536  6.051073 -0.243610   \n",
      "2  7.473760 -9.258449  5.053413 -4.237329  5.577164 -8.749300 -1.383737   \n",
      "4  0.094886  3.894744 -4.178989  4.780040  1.429170  1.864098 -3.345717   \n",
      "5 -1.356596  0.546319 -2.566127 -0.200193  0.049284 -1.638596  2.954736   \n",
      "\n",
      "      PCA_13    PCA_14  \n",
      "0  -3.168678 -1.712049  \n",
      "1   1.750005 -4.635485  \n",
      "2  12.193140  9.484669  \n",
      "4   0.923514  3.746379  \n",
      "5   5.486645  0.411145  \n",
      "{'lung_1': 0, 'lung_10': 1, 'lung_11': 2, 'lung_12': 3, 'lung_13': 4, 'lung_14': 5, 'lung_15': 6, 'lung_16': 7, 'lung_17': 8, 'lung_18': 9, 'lung_19': 10, 'lung_2': 11, 'lung_20': 12, 'lung_21': 13, 'lung_22': 14, 'lung_23': 15, 'lung_24': 16, 'lung_25': 17, 'lung_26': 18, 'lung_27': 19, 'lung_28': 20, 'lung_29': 21, 'lung_3': 22, 'lung_30': 23, 'lung_31': 24, 'lung_32': 25, 'lung_33': 26, 'lung_34': 27, 'lung_35': 28, 'lung_36': 29, 'lung_37': 30, 'lung_38': 31, 'lung_39': 32, 'lung_40': 33, 'lung_41': 34, 'lung_42': 35, 'lung_43': 36, 'lung_44': 37, 'lung_45': 38, 'lung_46': 39, 'lung_47': 40, 'lung_48': 41, 'lung_49': 42, 'lung_5': 43, 'lung_50': 44, 'lung_51': 45, 'lung_52': 46, 'lung_53': 47, 'lung_54': 48, 'lung_55': 49, 'lung_56': 50, 'lung_57': 51, 'lung_58': 52, 'lung_59': 53, 'lung_6': 54, 'lung_60': 55, 'lung_61': 56, 'lung_62': 57, 'lung_63': 58, 'lung_64': 59, 'lung_65': 60, 'lung_66': 61, 'lung_67': 62, 'lung_68': 63, 'lung_69': 64, 'lung_7': 65, 'lung_70': 66, 'lung_71': 67, 'lung_72': 68, 'lung_73': 69, 'lung_74': 70, 'lung_75': 71, 'lung_76': 72, 'lung_77': 73, 'lung_78': 74, 'lung_79': 75, 'lung_8': 76, 'lung_80': 77, 'lung_81': 78, 'lung_82': 79, 'lung_83': 80, 'lung_84': 81, 'lung_85': 82, 'lung_86': 83, 'lung_87': 84, 'lung_88': 85, 'lung_89': 86, 'lung_9': 87}\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<style>#sk-container-id-1 {\n",
       "  /* Definition of color scheme common for light and dark mode */\n",
       "  --sklearn-color-text: #000;\n",
       "  --sklearn-color-text-muted: #666;\n",
       "  --sklearn-color-line: gray;\n",
       "  /* Definition of color scheme for unfitted estimators */\n",
       "  --sklearn-color-unfitted-level-0: #fff5e6;\n",
       "  --sklearn-color-unfitted-level-1: #f6e4d2;\n",
       "  --sklearn-color-unfitted-level-2: #ffe0b3;\n",
       "  --sklearn-color-unfitted-level-3: chocolate;\n",
       "  /* Definition of color scheme for fitted estimators */\n",
       "  --sklearn-color-fitted-level-0: #f0f8ff;\n",
       "  --sklearn-color-fitted-level-1: #d4ebff;\n",
       "  --sklearn-color-fitted-level-2: #b3dbfd;\n",
       "  --sklearn-color-fitted-level-3: cornflowerblue;\n",
       "\n",
       "  /* Specific color for light theme */\n",
       "  --sklearn-color-text-on-default-background: var(--sg-text-color, var(--theme-code-foreground, var(--jp-content-font-color1, black)));\n",
       "  --sklearn-color-background: var(--sg-background-color, var(--theme-background, var(--jp-layout-color0, white)));\n",
       "  --sklearn-color-border-box: var(--sg-text-color, var(--theme-code-foreground, var(--jp-content-font-color1, black)));\n",
       "  --sklearn-color-icon: #696969;\n",
       "\n",
       "  @media (prefers-color-scheme: dark) {\n",
       "    /* Redefinition of color scheme for dark theme */\n",
       "    --sklearn-color-text-on-default-background: var(--sg-text-color, var(--theme-code-foreground, var(--jp-content-font-color1, white)));\n",
       "    --sklearn-color-background: var(--sg-background-color, var(--theme-background, var(--jp-layout-color0, #111)));\n",
       "    --sklearn-color-border-box: var(--sg-text-color, var(--theme-code-foreground, var(--jp-content-font-color1, white)));\n",
       "    --sklearn-color-icon: #878787;\n",
       "  }\n",
       "}\n",
       "\n",
       "#sk-container-id-1 {\n",
       "  color: var(--sklearn-color-text);\n",
       "}\n",
       "\n",
       "#sk-container-id-1 pre {\n",
       "  padding: 0;\n",
       "}\n",
       "\n",
       "#sk-container-id-1 input.sk-hidden--visually {\n",
       "  border: 0;\n",
       "  clip: rect(1px 1px 1px 1px);\n",
       "  clip: rect(1px, 1px, 1px, 1px);\n",
       "  height: 1px;\n",
       "  margin: -1px;\n",
       "  overflow: hidden;\n",
       "  padding: 0;\n",
       "  position: absolute;\n",
       "  width: 1px;\n",
       "}\n",
       "\n",
       "#sk-container-id-1 div.sk-dashed-wrapped {\n",
       "  border: 1px dashed var(--sklearn-color-line);\n",
       "  margin: 0 0.4em 0.5em 0.4em;\n",
       "  box-sizing: border-box;\n",
       "  padding-bottom: 0.4em;\n",
       "  background-color: var(--sklearn-color-background);\n",
       "}\n",
       "\n",
       "#sk-container-id-1 div.sk-container {\n",
       "  /* jupyter's `normalize.less` sets `[hidden] { display: none; }`\n",
       "     but bootstrap.min.css set `[hidden] { display: none !important; }`\n",
       "     so we also need the `!important` here to be able to override the\n",
       "     default hidden behavior on the sphinx rendered scikit-learn.org.\n",
       "     See: https://github.com/scikit-learn/scikit-learn/issues/21755 */\n",
       "  display: inline-block !important;\n",
       "  position: relative;\n",
       "}\n",
       "\n",
       "#sk-container-id-1 div.sk-text-repr-fallback {\n",
       "  display: none;\n",
       "}\n",
       "\n",
       "div.sk-parallel-item,\n",
       "div.sk-serial,\n",
       "div.sk-item {\n",
       "  /* draw centered vertical line to link estimators */\n",
       "  background-image: linear-gradient(var(--sklearn-color-text-on-default-background), var(--sklearn-color-text-on-default-background));\n",
       "  background-size: 2px 100%;\n",
       "  background-repeat: no-repeat;\n",
       "  background-position: center center;\n",
       "}\n",
       "\n",
       "/* Parallel-specific style estimator block */\n",
       "\n",
       "#sk-container-id-1 div.sk-parallel-item::after {\n",
       "  content: \"\";\n",
       "  width: 100%;\n",
       "  border-bottom: 2px solid var(--sklearn-color-text-on-default-background);\n",
       "  flex-grow: 1;\n",
       "}\n",
       "\n",
       "#sk-container-id-1 div.sk-parallel {\n",
       "  display: flex;\n",
       "  align-items: stretch;\n",
       "  justify-content: center;\n",
       "  background-color: var(--sklearn-color-background);\n",
       "  position: relative;\n",
       "}\n",
       "\n",
       "#sk-container-id-1 div.sk-parallel-item {\n",
       "  display: flex;\n",
       "  flex-direction: column;\n",
       "}\n",
       "\n",
       "#sk-container-id-1 div.sk-parallel-item:first-child::after {\n",
       "  align-self: flex-end;\n",
       "  width: 50%;\n",
       "}\n",
       "\n",
       "#sk-container-id-1 div.sk-parallel-item:last-child::after {\n",
       "  align-self: flex-start;\n",
       "  width: 50%;\n",
       "}\n",
       "\n",
       "#sk-container-id-1 div.sk-parallel-item:only-child::after {\n",
       "  width: 0;\n",
       "}\n",
       "\n",
       "/* Serial-specific style estimator block */\n",
       "\n",
       "#sk-container-id-1 div.sk-serial {\n",
       "  display: flex;\n",
       "  flex-direction: column;\n",
       "  align-items: center;\n",
       "  background-color: var(--sklearn-color-background);\n",
       "  padding-right: 1em;\n",
       "  padding-left: 1em;\n",
       "}\n",
       "\n",
       "\n",
       "/* Toggleable style: style used for estimator/Pipeline/ColumnTransformer box that is\n",
       "clickable and can be expanded/collapsed.\n",
       "- Pipeline and ColumnTransformer use this feature and define the default style\n",
       "- Estimators will overwrite some part of the style using the `sk-estimator` class\n",
       "*/\n",
       "\n",
       "/* Pipeline and ColumnTransformer style (default) */\n",
       "\n",
       "#sk-container-id-1 div.sk-toggleable {\n",
       "  /* Default theme specific background. It is overwritten whether we have a\n",
       "  specific estimator or a Pipeline/ColumnTransformer */\n",
       "  background-color: var(--sklearn-color-background);\n",
       "}\n",
       "\n",
       "/* Toggleable label */\n",
       "#sk-container-id-1 label.sk-toggleable__label {\n",
       "  cursor: pointer;\n",
       "  display: flex;\n",
       "  width: 100%;\n",
       "  margin-bottom: 0;\n",
       "  padding: 0.5em;\n",
       "  box-sizing: border-box;\n",
       "  text-align: center;\n",
       "  align-items: start;\n",
       "  justify-content: space-between;\n",
       "  gap: 0.5em;\n",
       "}\n",
       "\n",
       "#sk-container-id-1 label.sk-toggleable__label .caption {\n",
       "  font-size: 0.6rem;\n",
       "  font-weight: lighter;\n",
       "  color: var(--sklearn-color-text-muted);\n",
       "}\n",
       "\n",
       "#sk-container-id-1 label.sk-toggleable__label-arrow:before {\n",
       "  /* Arrow on the left of the label */\n",
       "  content: \"▸\";\n",
       "  float: left;\n",
       "  margin-right: 0.25em;\n",
       "  color: var(--sklearn-color-icon);\n",
       "}\n",
       "\n",
       "#sk-container-id-1 label.sk-toggleable__label-arrow:hover:before {\n",
       "  color: var(--sklearn-color-text);\n",
       "}\n",
       "\n",
       "/* Toggleable content - dropdown */\n",
       "\n",
       "#sk-container-id-1 div.sk-toggleable__content {\n",
       "  display: none;\n",
       "  text-align: left;\n",
       "  /* unfitted */\n",
       "  background-color: var(--sklearn-color-unfitted-level-0);\n",
       "}\n",
       "\n",
       "#sk-container-id-1 div.sk-toggleable__content.fitted {\n",
       "  /* fitted */\n",
       "  background-color: var(--sklearn-color-fitted-level-0);\n",
       "}\n",
       "\n",
       "#sk-container-id-1 div.sk-toggleable__content pre {\n",
       "  margin: 0.2em;\n",
       "  border-radius: 0.25em;\n",
       "  color: var(--sklearn-color-text);\n",
       "  /* unfitted */\n",
       "  background-color: var(--sklearn-color-unfitted-level-0);\n",
       "}\n",
       "\n",
       "#sk-container-id-1 div.sk-toggleable__content.fitted pre {\n",
       "  /* unfitted */\n",
       "  background-color: var(--sklearn-color-fitted-level-0);\n",
       "}\n",
       "\n",
       "#sk-container-id-1 input.sk-toggleable__control:checked~div.sk-toggleable__content {\n",
       "  /* Expand drop-down */\n",
       "  display: block;\n",
       "  width: 100%;\n",
       "  overflow: visible;\n",
       "}\n",
       "\n",
       "#sk-container-id-1 input.sk-toggleable__control:checked~label.sk-toggleable__label-arrow:before {\n",
       "  content: \"▾\";\n",
       "}\n",
       "\n",
       "/* Pipeline/ColumnTransformer-specific style */\n",
       "\n",
       "#sk-container-id-1 div.sk-label input.sk-toggleable__control:checked~label.sk-toggleable__label {\n",
       "  color: var(--sklearn-color-text);\n",
       "  background-color: var(--sklearn-color-unfitted-level-2);\n",
       "}\n",
       "\n",
       "#sk-container-id-1 div.sk-label.fitted input.sk-toggleable__control:checked~label.sk-toggleable__label {\n",
       "  background-color: var(--sklearn-color-fitted-level-2);\n",
       "}\n",
       "\n",
       "/* Estimator-specific style */\n",
       "\n",
       "/* Colorize estimator box */\n",
       "#sk-container-id-1 div.sk-estimator input.sk-toggleable__control:checked~label.sk-toggleable__label {\n",
       "  /* unfitted */\n",
       "  background-color: var(--sklearn-color-unfitted-level-2);\n",
       "}\n",
       "\n",
       "#sk-container-id-1 div.sk-estimator.fitted input.sk-toggleable__control:checked~label.sk-toggleable__label {\n",
       "  /* fitted */\n",
       "  background-color: var(--sklearn-color-fitted-level-2);\n",
       "}\n",
       "\n",
       "#sk-container-id-1 div.sk-label label.sk-toggleable__label,\n",
       "#sk-container-id-1 div.sk-label label {\n",
       "  /* The background is the default theme color */\n",
       "  color: var(--sklearn-color-text-on-default-background);\n",
       "}\n",
       "\n",
       "/* On hover, darken the color of the background */\n",
       "#sk-container-id-1 div.sk-label:hover label.sk-toggleable__label {\n",
       "  color: var(--sklearn-color-text);\n",
       "  background-color: var(--sklearn-color-unfitted-level-2);\n",
       "}\n",
       "\n",
       "/* Label box, darken color on hover, fitted */\n",
       "#sk-container-id-1 div.sk-label.fitted:hover label.sk-toggleable__label.fitted {\n",
       "  color: var(--sklearn-color-text);\n",
       "  background-color: var(--sklearn-color-fitted-level-2);\n",
       "}\n",
       "\n",
       "/* Estimator label */\n",
       "\n",
       "#sk-container-id-1 div.sk-label label {\n",
       "  font-family: monospace;\n",
       "  font-weight: bold;\n",
       "  display: inline-block;\n",
       "  line-height: 1.2em;\n",
       "}\n",
       "\n",
       "#sk-container-id-1 div.sk-label-container {\n",
       "  text-align: center;\n",
       "}\n",
       "\n",
       "/* Estimator-specific */\n",
       "#sk-container-id-1 div.sk-estimator {\n",
       "  font-family: monospace;\n",
       "  border: 1px dotted var(--sklearn-color-border-box);\n",
       "  border-radius: 0.25em;\n",
       "  box-sizing: border-box;\n",
       "  margin-bottom: 0.5em;\n",
       "  /* unfitted */\n",
       "  background-color: var(--sklearn-color-unfitted-level-0);\n",
       "}\n",
       "\n",
       "#sk-container-id-1 div.sk-estimator.fitted {\n",
       "  /* fitted */\n",
       "  background-color: var(--sklearn-color-fitted-level-0);\n",
       "}\n",
       "\n",
       "/* on hover */\n",
       "#sk-container-id-1 div.sk-estimator:hover {\n",
       "  /* unfitted */\n",
       "  background-color: var(--sklearn-color-unfitted-level-2);\n",
       "}\n",
       "\n",
       "#sk-container-id-1 div.sk-estimator.fitted:hover {\n",
       "  /* fitted */\n",
       "  background-color: var(--sklearn-color-fitted-level-2);\n",
       "}\n",
       "\n",
       "/* Specification for estimator info (e.g. \"i\" and \"?\") */\n",
       "\n",
       "/* Common style for \"i\" and \"?\" */\n",
       "\n",
       ".sk-estimator-doc-link,\n",
       "a:link.sk-estimator-doc-link,\n",
       "a:visited.sk-estimator-doc-link {\n",
       "  float: right;\n",
       "  font-size: smaller;\n",
       "  line-height: 1em;\n",
       "  font-family: monospace;\n",
       "  background-color: var(--sklearn-color-background);\n",
       "  border-radius: 1em;\n",
       "  height: 1em;\n",
       "  width: 1em;\n",
       "  text-decoration: none !important;\n",
       "  margin-left: 0.5em;\n",
       "  text-align: center;\n",
       "  /* unfitted */\n",
       "  border: var(--sklearn-color-unfitted-level-1) 1pt solid;\n",
       "  color: var(--sklearn-color-unfitted-level-1);\n",
       "}\n",
       "\n",
       ".sk-estimator-doc-link.fitted,\n",
       "a:link.sk-estimator-doc-link.fitted,\n",
       "a:visited.sk-estimator-doc-link.fitted {\n",
       "  /* fitted */\n",
       "  border: var(--sklearn-color-fitted-level-1) 1pt solid;\n",
       "  color: var(--sklearn-color-fitted-level-1);\n",
       "}\n",
       "\n",
       "/* On hover */\n",
       "div.sk-estimator:hover .sk-estimator-doc-link:hover,\n",
       ".sk-estimator-doc-link:hover,\n",
       "div.sk-label-container:hover .sk-estimator-doc-link:hover,\n",
       ".sk-estimator-doc-link:hover {\n",
       "  /* unfitted */\n",
       "  background-color: var(--sklearn-color-unfitted-level-3);\n",
       "  color: var(--sklearn-color-background);\n",
       "  text-decoration: none;\n",
       "}\n",
       "\n",
       "div.sk-estimator.fitted:hover .sk-estimator-doc-link.fitted:hover,\n",
       ".sk-estimator-doc-link.fitted:hover,\n",
       "div.sk-label-container:hover .sk-estimator-doc-link.fitted:hover,\n",
       ".sk-estimator-doc-link.fitted:hover {\n",
       "  /* fitted */\n",
       "  background-color: var(--sklearn-color-fitted-level-3);\n",
       "  color: var(--sklearn-color-background);\n",
       "  text-decoration: none;\n",
       "}\n",
       "\n",
       "/* Span, style for the box shown on hovering the info icon */\n",
       ".sk-estimator-doc-link span {\n",
       "  display: none;\n",
       "  z-index: 9999;\n",
       "  position: relative;\n",
       "  font-weight: normal;\n",
       "  right: .2ex;\n",
       "  padding: .5ex;\n",
       "  margin: .5ex;\n",
       "  width: min-content;\n",
       "  min-width: 20ex;\n",
       "  max-width: 50ex;\n",
       "  color: var(--sklearn-color-text);\n",
       "  box-shadow: 2pt 2pt 4pt #999;\n",
       "  /* unfitted */\n",
       "  background: var(--sklearn-color-unfitted-level-0);\n",
       "  border: .5pt solid var(--sklearn-color-unfitted-level-3);\n",
       "}\n",
       "\n",
       ".sk-estimator-doc-link.fitted span {\n",
       "  /* fitted */\n",
       "  background: var(--sklearn-color-fitted-level-0);\n",
       "  border: var(--sklearn-color-fitted-level-3);\n",
       "}\n",
       "\n",
       ".sk-estimator-doc-link:hover span {\n",
       "  display: block;\n",
       "}\n",
       "\n",
       "/* \"?\"-specific style due to the `<a>` HTML tag */\n",
       "\n",
       "#sk-container-id-1 a.estimator_doc_link {\n",
       "  float: right;\n",
       "  font-size: 1rem;\n",
       "  line-height: 1em;\n",
       "  font-family: monospace;\n",
       "  background-color: var(--sklearn-color-background);\n",
       "  border-radius: 1rem;\n",
       "  height: 1rem;\n",
       "  width: 1rem;\n",
       "  text-decoration: none;\n",
       "  /* unfitted */\n",
       "  color: var(--sklearn-color-unfitted-level-1);\n",
       "  border: var(--sklearn-color-unfitted-level-1) 1pt solid;\n",
       "}\n",
       "\n",
       "#sk-container-id-1 a.estimator_doc_link.fitted {\n",
       "  /* fitted */\n",
       "  border: var(--sklearn-color-fitted-level-1) 1pt solid;\n",
       "  color: var(--sklearn-color-fitted-level-1);\n",
       "}\n",
       "\n",
       "/* On hover */\n",
       "#sk-container-id-1 a.estimator_doc_link:hover {\n",
       "  /* unfitted */\n",
       "  background-color: var(--sklearn-color-unfitted-level-3);\n",
       "  color: var(--sklearn-color-background);\n",
       "  text-decoration: none;\n",
       "}\n",
       "\n",
       "#sk-container-id-1 a.estimator_doc_link.fitted:hover {\n",
       "  /* fitted */\n",
       "  background-color: var(--sklearn-color-fitted-level-3);\n",
       "}\n",
       "\n",
       ".estimator-table summary {\n",
       "    padding: .5rem;\n",
       "    font-family: monospace;\n",
       "    cursor: pointer;\n",
       "}\n",
       "\n",
       ".estimator-table details[open] {\n",
       "    padding-left: 0.1rem;\n",
       "    padding-right: 0.1rem;\n",
       "    padding-bottom: 0.3rem;\n",
       "}\n",
       "\n",
       ".estimator-table .parameters-table {\n",
       "    margin-left: auto !important;\n",
       "    margin-right: auto !important;\n",
       "}\n",
       "\n",
       ".estimator-table .parameters-table tr:nth-child(odd) {\n",
       "    background-color: #fff;\n",
       "}\n",
       "\n",
       ".estimator-table .parameters-table tr:nth-child(even) {\n",
       "    background-color: #f6f6f6;\n",
       "}\n",
       "\n",
       ".estimator-table .parameters-table tr:hover {\n",
       "    background-color: #e0e0e0;\n",
       "}\n",
       "\n",
       ".estimator-table table td {\n",
       "    border: 1px solid rgba(106, 105, 104, 0.232);\n",
       "}\n",
       "\n",
       ".user-set td {\n",
       "    color:rgb(255, 94, 0);\n",
       "    text-align: left;\n",
       "}\n",
       "\n",
       ".user-set td.value pre {\n",
       "    color:rgb(255, 94, 0) !important;\n",
       "    background-color: transparent !important;\n",
       "}\n",
       "\n",
       ".default td {\n",
       "    color: black;\n",
       "    text-align: left;\n",
       "}\n",
       "\n",
       ".user-set td i,\n",
       ".default td i {\n",
       "    color: black;\n",
       "}\n",
       "\n",
       ".copy-paste-icon {\n",
       "    background-image: url(data:image/svg+xml;base64,PHN2ZyB4bWxucz0iaHR0cDovL3d3dy53My5vcmcvMjAwMC9zdmciIHZpZXdCb3g9IjAgMCA0NDggNTEyIj48IS0tIUZvbnQgQXdlc29tZSBGcmVlIDYuNy4yIGJ5IEBmb250YXdlc29tZSAtIGh0dHBzOi8vZm9udGF3ZXNvbWUuY29tIExpY2Vuc2UgLSBodHRwczovL2ZvbnRhd2Vzb21lLmNvbS9saWNlbnNlL2ZyZWUgQ29weXJpZ2h0IDIwMjUgRm9udGljb25zLCBJbmMuLS0+PHBhdGggZD0iTTIwOCAwTDMzMi4xIDBjMTIuNyAwIDI0LjkgNS4xIDMzLjkgMTQuMWw2Ny45IDY3LjljOSA5IDE0LjEgMjEuMiAxNC4xIDMzLjlMNDQ4IDMzNmMwIDI2LjUtMjEuNSA0OC00OCA0OGwtMTkyIDBjLTI2LjUgMC00OC0yMS41LTQ4LTQ4bDAtMjg4YzAtMjYuNSAyMS41LTQ4IDQ4LTQ4ek00OCAxMjhsODAgMCAwIDY0LTY0IDAgMCAyNTYgMTkyIDAgMC0zMiA2NCAwIDAgNDhjMCAyNi41LTIxLjUgNDgtNDggNDhMNDggNTEyYy0yNi41IDAtNDgtMjEuNS00OC00OEwwIDE3NmMwLTI2LjUgMjEuNS00OCA0OC00OHoiLz48L3N2Zz4=);\n",
       "    background-repeat: no-repeat;\n",
       "    background-size: 14px 14px;\n",
       "    background-position: 0;\n",
       "    display: inline-block;\n",
       "    width: 14px;\n",
       "    height: 14px;\n",
       "    cursor: pointer;\n",
       "}\n",
       "</style><body><div id=\"sk-container-id-1\" class=\"sk-top-container\"><div class=\"sk-text-repr-fallback\"><pre>LabelEncoder()</pre><b>In a Jupyter environment, please rerun this cell to show the HTML representation or trust the notebook. <br />On GitHub, the HTML representation is unable to render, please try loading this page with nbviewer.org.</b></div><div class=\"sk-container\" hidden><div class=\"sk-item\"><div class=\"sk-estimator fitted sk-toggleable\"><input class=\"sk-toggleable__control sk-hidden--visually\" id=\"sk-estimator-id-1\" type=\"checkbox\" checked><label for=\"sk-estimator-id-1\" class=\"sk-toggleable__label fitted sk-toggleable__label-arrow\"><div><div>LabelEncoder</div></div><div><a class=\"sk-estimator-doc-link fitted\" rel=\"noreferrer\" target=\"_blank\" href=\"https://scikit-learn.org/1.7/modules/generated/sklearn.preprocessing.LabelEncoder.html\">?<span>Documentation for LabelEncoder</span></a><span class=\"sk-estimator-doc-link fitted\">i<span>Fitted</span></span></div></label><div class=\"sk-toggleable__content fitted\" data-param-prefix=\"\">\n",
       "        <div class=\"estimator-table\">\n",
       "            <details>\n",
       "                <summary>Parameters</summary>\n",
       "                <table class=\"parameters-table\">\n",
       "                  <tbody>\n",
       "                    \n",
       "                  </tbody>\n",
       "                </table>\n",
       "            </details>\n",
       "        </div>\n",
       "    </div></div></div></div></div><script>function copyToClipboard(text, element) {\n",
       "    // Get the parameter prefix from the closest toggleable content\n",
       "    const toggleableContent = element.closest('.sk-toggleable__content');\n",
       "    const paramPrefix = toggleableContent ? toggleableContent.dataset.paramPrefix : '';\n",
       "    const fullParamName = paramPrefix ? `${paramPrefix}${text}` : text;\n",
       "\n",
       "    const originalStyle = element.style;\n",
       "    const computedStyle = window.getComputedStyle(element);\n",
       "    const originalWidth = computedStyle.width;\n",
       "    const originalHTML = element.innerHTML.replace('Copied!', '');\n",
       "\n",
       "    navigator.clipboard.writeText(fullParamName)\n",
       "        .then(() => {\n",
       "            element.style.width = originalWidth;\n",
       "            element.style.color = 'green';\n",
       "            element.innerHTML = \"Copied!\";\n",
       "\n",
       "            setTimeout(() => {\n",
       "                element.innerHTML = originalHTML;\n",
       "                element.style = originalStyle;\n",
       "            }, 2000);\n",
       "        })\n",
       "        .catch(err => {\n",
       "            console.error('Failed to copy:', err);\n",
       "            element.style.color = 'red';\n",
       "            element.innerHTML = \"Failed!\";\n",
       "            setTimeout(() => {\n",
       "                element.innerHTML = originalHTML;\n",
       "                element.style = originalStyle;\n",
       "            }, 2000);\n",
       "        });\n",
       "    return false;\n",
       "}\n",
       "\n",
       "document.querySelectorAll('.fa-regular.fa-copy').forEach(function(element) {\n",
       "    const toggleableContent = element.closest('.sk-toggleable__content');\n",
       "    const paramPrefix = toggleableContent ? toggleableContent.dataset.paramPrefix : '';\n",
       "    const paramName = element.parentElement.nextElementSibling.textContent.trim();\n",
       "    const fullParamName = paramPrefix ? `${paramPrefix}${paramName}` : paramName;\n",
       "\n",
       "    element.setAttribute('title', fullParamName);\n",
       "});\n",
       "</script></body>"
      ],
      "text/plain": [
       "LabelEncoder()"
      ]
     },
     "execution_count": 1,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "from DataProcessor import DataProcessor\n",
    "\n",
    "pipeline = DataProcessor(input_file_exp = 'GSE58661_series_matrix.txt.gz', input_file_meta = \"Lung3.metadata.xls\",\n",
    "                         cols_to_grab = ['title',\n",
    "       'characteristics.tag.gender', 'characteristics.tag.histology',\n",
    "       'characteristics.tag.tumor.size.maximumdiameter',\n",
    "       'characteristics.tag.stage.primary.tumor'])\n",
    "\n",
    "pipeline_finished, encoders = pipeline.pipeline_run()\n",
    "\n",
    "print(pipeline_finished.head())\n",
    "pipeline.look_at_encode(encoders, \"title\")\n",
    "\n",
    "\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "<a name='03'></a>\n",
    "# Part C. Model Evaluation\n",
    "\n",
    "Evaluating the performance of a machine learning model goes beyond just looking at accuracy, as accuracy alone can be misleading, especially in cases where the dataset is imbalanced or where different types of errors have different consequences\n",
    "- **Detecting Overfitting/Underfitting**: A learning curve can help you understand whether your model is overfitting (performing well on training data but poorly on validation data) or underfitting (performing poorly on both training and validation data)\n",
    "- **ROC**: The ROC curve shows how well your model distinguishes between classes. It helps in selecting the optimal threshold for classification decisions, particularly when the cost of false positives and false negatives differs significantly. ROC curves are often used to compare models. \n",
    "- **Confusion matrix**: From the confusion matrix, you can derive other important metrics like precision, recall, F1-score, and specificity, which give a better understanding of how your model is performing across different classes\n",
    "\n",
    "\n",
    "### <span style=\"background-color: lightyellow;\">Evaluation Task</span>\n",
    "- Review the documentation for `sklearn.metrics` and `learning_curve`.\n",
    "- Select appropriate metrics for your dataset, including accuracy and indicators of overfitting or underfitting.\n",
    "- Implement functions to compute and assess these metrics. It is allowed to use libraries.\n",
    "- Add the evaluation functions as a Python module to your repository.\n",
    "- Update the evaluation documentation to clearly explain your choices for the evaluation metrics and the rationale behind them.\n",
    "\n",
    "See also: [Model evaluation video](https://video.hanze.nl/media/model-evaluation/0_gybpnhq7)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# YOUR CODE HERE TO DEMONSTRATE THE USAGE OF THE EVALUATION MODULE"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "<a name='04'></a>\n",
    "## Part D. Ensemble learning\n",
    "Ensemble learning is a powerful machine learning technique that combines the predictions of multiple models to improve overall performance, robustness, and accuracy. Rather than relying on a single model, ensemble learning methods aggregate the results of several models—often called \"weak learners\"—to produce a stronger predictive model. The key idea behind ensemble learning is that by combining models, the weaknesses of individual models can be offset, leading to better generalization on unseen data. Popular ensemble ML algorithms are the `Random Forest` and `XGBoost`. \n",
    "Here’s an improved version of the introduction:\n",
    "\n",
    "### voting algorithms\n",
    "Voting algorithms in ensemble learning combine the predictions of multiple classifiers to make a final decision, typically based on the consensus or weighted agreement among models. Two primary types of voting are commonly used: hard voting, where the final prediction is determined by the majority vote, and soft voting, which uses the weighted average of predicted probabilities to determine the outcome. Several sites explain the hard and soft voting algorithm. A clear explanation can be found on https://www.baeldung.com/cs/hard-vs-soft-voting-classifiers\n",
    "\n",
    "\n",
    "### <span style=\"background-color: lightyellow;\">Ensemble Task</span>\n",
    "- Review the [Study Case notebook on ensemble learning](..Study_Cases/study_case_bagging_boosting.ipynb).\n",
    "- Try three different algorithms for classification of your data label.\n",
    "- Implement a hard and soft voting algorithm for model aggregation.\n",
    "- Compare the performance of the voting algorithm with that of a boosting or bagging algorithm.\n",
    "- Update your repository with the voting algorithm class\n",
    "\n",
    "See also: [ensemle learning video](https://video.hanze.nl/media/Ensemble/0_sue5v33g)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# YOUR CODE HERE WITH THE VOTING ALGORITHMS\n",
    "def vote_soft():\n",
    "    pass\n",
    "\n",
    "def vote_hard():\n",
    "    pass"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# YOUR CODE HERE TO COMPARE VOTING OUTCOME WITH A BAGGING OR BOOSTING ALGORITHM FROM SKLEARN"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "<a name='05'></a>\n",
    "## Part E. Pipelines\n",
    "The sklearn pipeline function is a tool in machine learning that simplifies the workflow by encapsulating all the steps involved in a single object. It offers advantages such as simplicity, reproducibility, efficiency, flexibility, and integration. The Pipeline is built using a list of (key, value) pairs, where the key is a string containing the name you want to give this step and value is an estimator object (the method to be executed).\n",
    "\n",
    "### <span style=\"background-color: lightyellow;\">Pipeline Task</span>\n",
    "- Read the [Study Case notebook for a pipeline functions](..Study_Cases/study_case_pipeline.ipynb) to understand the principle of the pipeline function\n",
    "- Implement a pipeline which prepares and classifies data \n",
    "- Use a `GridSearchCV` object with the `Pipeline` object and a parameter grid to optimize choose the best hyper parameters"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# YOUR CODE HERE WITH THE PIPELINE IMPLEMENTATION"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## <span style=\"background-color: lightyellow;\">Bring it all together</span>\n",
    "\n",
    "By now, you've developed code snippets for model evaluation, optimization, and data improvement. Now, leverage these skills to build a classification model using the Clinical and Genetic Lung data. Make sure that you log your experiments. \n",
    "\n",
    "Once you're satisfied with the model, upload the relevant code to your repository and or refactor code with new insights. Furthermore, take a moment to reflect on its applicability and potential real-world impact. Update your evaluation document(s) with these findings in your repository. "
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Bonus: ML for operations\n",
    "\n",
    "If we intend to deploy the model in a real-world application, it's more efficient to save and reuse the trained model rather than retraining it each time. \n",
    "\n",
    "- Read the blog: https://neptune.ai/blog/saving-trained-model-in-python\n",
    "- Write three python files \n",
    "    1) a train_model python file \n",
    "    2) a use_model python file\n",
    "    3) a retrain_model python file that adds new data to the original training data and updates the model\n",
    "- Update your repository\n"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "dsc5",
   "language": "python",
   "name": "dsc5"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.13.9"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 4
}
